<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>taxreturn-vignette</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">taxreturn-vignette</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">library</span>(taxreturn)</a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw">library</span>(Biostrings)</a>
<a class="sourceLine" id="cb1-3" title="3"><span class="kw">library</span>(tidyverse)</a></code></pre></div>
<div id="fetching-sequences-from-genbank-and-bold" class="section level2">
<h2>Fetching sequences from GenBank and BOLD</h2>
<p>The first step is to retrieve public reference data sequences from NCBI Genbank and BOLD. This can be done with the fetchSeqs function, which wraps an interface to entrez and BOLD API’s respectively. This function can either take a single higher level rank, i.e. the family ‘Trioza’, or it can take a vector of taxon names, such as the species contained on a priority pest list or those of conservation concern.</p>
<p>The downstream option lets the function know if you want to conduct searches and output fasta files using the input rank, or at a taxonomic rank downstream of it, i.e. Species. The downto option then determines what that downstream rank is. This functionality is important for getting around server limits when conducting large searches for example ‘Insecta’, however comes with some computational overhead when downloading few taxa.</p>
<p>The marker option determines what will be in the search, note that the naming of loci differs between Genbank and bold so i suggest conducting a test search on their respective websites to confirm the desired marker. <em>note - add functionality for checking if marker exists</em></p>
<p>The compress option toggles whether to output as a gzipped fasta file to save space. All further functions in this package should be capable of handling gzipped files.</p>
<p>Finally, the cores option determines how many cores to use when downloading sequences, using more cores can speed up searches. <em>note - i dont believe cores is currently working properly so use 1 core for now</em></p>
<p>Depending on the amount of sequences you are downloading, and the speed of your internet connection this step can take from minutes to hours, i suggest running this overnight for large searches.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="co">## Fetch sequences from GenBank </span></a>
<a class="sourceLine" id="cb2-2" title="2">genbank &lt;-<span class="st"> </span><span class="kw">fetchSeqs</span>(<span class="st">&quot;Scaptodrosophila&quot;</span>, <span class="dt">database=</span><span class="st">&quot;genbank&quot;</span>,<span class="dt">dir=</span><span class="st">&quot;genbank&quot;</span>,<span class="dt">downstream=</span><span class="ot">TRUE</span>,<span class="dt">quiet=</span><span class="ot">FALSE</span>, <span class="dt">downto=</span><span class="st">&quot;species&quot;</span>, <span class="dt">marker=</span><span class="st">&quot;COI OR COI OR COX1 OR COXI&quot;</span>, <span class="dt">output =</span> <span class="st">&quot;gb-binom&quot;</span>,<span class="dt">compress=</span><span class="ot">TRUE</span>, <span class="dt">cores=</span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb2-3" title="3"></a>
<a class="sourceLine" id="cb2-4" title="4"></a>
<a class="sourceLine" id="cb2-5" title="5"><span class="co">## Fetch sequences from BOLD</span></a>
<a class="sourceLine" id="cb2-6" title="6">bold &lt;-<span class="st"> </span><span class="kw">fetchSeqs</span>(<span class="st">&quot;Scaptodrosophila&quot;</span>, <span class="dt">database=</span><span class="st">&quot;bold&quot;</span>, <span class="dt">dir=</span><span class="st">&quot;bold&quot;</span>, <span class="dt">downstream=</span><span class="ot">TRUE</span>,<span class="dt">quiet=</span><span class="ot">FALSE</span>, <span class="dt">downto=</span><span class="st">&quot;species&quot;</span>, <span class="dt">marker=</span><span class="st">&quot;COI-5P&quot;</span>, <span class="dt">output =</span> <span class="st">&quot;gb-binom&quot;</span>,<span class="dt">compress=</span><span class="ot">TRUE</span>, <span class="dt">cores=</span><span class="dv">1</span>)</a></code></pre></div>
</div>
<div id="curating-public-reference-sequences" class="section level1">
<h1>Curating public reference sequences</h1>
<div id="removing-non-homologous-sequences" class="section level2">
<h2>Removing non-homologous sequences</h2>
<p>Due to the prevalence of misannotated data on public reference data we will use a number of filtering steps to try and curate this to only the markers desired. The first curation step we wish to use is to remove non-homologous markers</p>
<p>In order to reduce bias involved in mapping sequences to a single reference sequence, we will align them to a profile hidden markov model of the gene that uses a probablistic framework to take into account a wider range of diversity. To make this model we will use the aphid R package, and a curated alignment of insect COI sequences obtained from the Midori dataset and trimmed to the folmer region.</p>
<p>This model can be loaded from the package data as below:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="co">#model &lt;- data(&quot;model&quot;, package=&quot;taxreturn&quot;)</span></a>
<a class="sourceLine" id="cb3-2" title="2"><span class="kw">load</span>(<span class="st">&quot;C:/Users/ap0y/Dropbox/R/taxreturn/data/model.rda&quot;</span>)</a></code></pre></div>
<p>Or a new model can be trained on a new dataset or loci using:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="co">#build PHMM from midori longest - sequences need to be same length</span></a>
<a class="sourceLine" id="cb4-2" title="2">midori &lt;-<span class="st">  </span>Biostrings<span class="op">::</span><span class="kw">readDNAStringSet</span>(<span class="st">&quot;MIDORI_LONGEST_20180221_COI.fasta&quot;</span>)</a>
<a class="sourceLine" id="cb4-3" title="3">insecta_midori &lt;-<span class="st"> </span><span class="kw">as.DNAbin</span>(midori[<span class="kw">str_detect</span>(<span class="kw">names</span>(midori),<span class="dt">pattern=</span><span class="st">&quot;;Insecta;&quot;</span>),])</a>
<a class="sourceLine" id="cb4-4" title="4">folmer &lt;-<span class="st"> </span>insect<span class="op">::</span><span class="kw">virtualPCR</span>(insecta_midori, <span class="dt">up =</span> <span class="st">&quot;TITCIACIAAYCAYAARGAYATTGG&quot;</span>,<span class="dt">down=</span> <span class="st">&quot;TAIACYTCIGGRTGICCRAARAAYCA&quot;</span>,<span class="dt">cores=</span><span class="dv">2</span>, <span class="dt">rcdown =</span> <span class="ot">TRUE</span>, <span class="dt">trimprimers =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb4-5" title="5">filt &lt;-<span class="st"> </span>folmer[<span class="kw">lengths</span>(folmer)<span class="op">==</span><span class="dv">658</span>]</a>
<a class="sourceLine" id="cb4-6" title="6"></a>
<a class="sourceLine" id="cb4-7" title="7"><span class="co">#alignment was then manually curated in geneious prime</span></a>
<a class="sourceLine" id="cb4-8" title="8">folmer_curated &lt;-<span class="st">  </span>ape<span class="op">::</span><span class="kw">read.dna</span>(<span class="st">&quot;folmer_insecta_fullength_aligned_curated.fa&quot;</span>,<span class="dt">format=</span><span class="st">&quot;fasta&quot;</span>)</a>
<a class="sourceLine" id="cb4-9" title="9">model &lt;-<span class="st"> </span>aphid<span class="op">::</span><span class="kw">derivePHMM</span>(folmer_curated)</a></code></pre></div>
<p>We will now use the cleanseqs function, and this model in order to remove putatively non-homolgous sequences. Firstly we will merge the sequences from each database together. As BOLD and GenBank share a number of sequences, we subset the merged file to only the unique sequences to speed things up.</p>
<p>As we only wish to look at the sequence data contained within the range of our alignment model (in this case the folmer region of COI), we use the option shave=TRUE to remove all bases to the left and right of the model.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="co">#read in all fastas and merge</span></a>
<a class="sourceLine" id="cb5-2" title="2"><span class="kw">library</span>(Biostrings)</a>
<a class="sourceLine" id="cb5-3" title="3">gbSeqs &lt;-<span class="st">  </span><span class="kw">readDNAStringSet</span>(<span class="kw">sort</span>(<span class="kw">list.files</span>(<span class="st">&quot;genbank&quot;</span>, <span class="dt">pattern =</span> <span class="st">&quot;.fa&quot;</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)))</a>
<a class="sourceLine" id="cb5-4" title="4">boldSeqs &lt;-<span class="st">  </span><span class="kw">readDNAStringSet</span>(<span class="kw">sort</span>(<span class="kw">list.files</span>(<span class="st">&quot;bold&quot;</span>, <span class="dt">pattern =</span> <span class="st">&quot;.fa&quot;</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)))</a>
<a class="sourceLine" id="cb5-5" title="5">mergedSeqs &lt;-<span class="st"> </span><span class="kw">append</span>(gbSeqs, boldSeqs, <span class="dt">after=</span><span class="kw">length</span>(gbSeqs))</a>
<a class="sourceLine" id="cb5-6" title="6">uniqSeqs &lt;-<span class="st"> </span>mergedSeqs[<span class="kw">unique</span>(<span class="kw">names</span>(mergedSeqs)),] <span class="co"># Remove those sequnce names that are identical across both databases</span></a>
<a class="sourceLine" id="cb5-7" title="7"></a>
<a class="sourceLine" id="cb5-8" title="8"><span class="co">#remove non-homologous sequences</span></a>
<a class="sourceLine" id="cb5-9" title="9">filtered &lt;-<span class="st"> </span><span class="kw">clean_seqs</span>(uniqSeqs, model, <span class="dt">minscore =</span> <span class="dv">100</span>, <span class="dt">cores=</span><span class="dv">2</span>, <span class="dt">shave=</span><span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="resolve-contaminated-sequences-and-misannotated-taxonomy" class="section level2">
<h2>Resolve Contaminated sequences and misannotated taxonomy</h2>
<p>The other main form of misannotation that can effect metabarcoding datasets is incorrect taxonomy for the reference sequences. To resolve this issue, we use the purge function from the insect R package, which clusters sequences at a specific similarity threshold (in this case 99% simularity), and compares the heirarchial taxonomy within clusters. When the taxonomy of a sequences diverges from the other sequences in its cluster, it is removed as a putative misannotation. The confidence required the remove a sequence can be adjusted. In this case we use a confidence threshold of 0.8, which indicates the putative misannotated sequence must diverge from 4/5 other sequences in its cluster to be removed from the dataset.</p>
<p>insect::purge requires specifically formated names, so we do some transformations to get the names in this format, retaining the old names in the attributes. These old names are then restored following removal of missanotated sequences</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="co">#Download the NCBI taxonomy database</span></a>
<a class="sourceLine" id="cb6-2" title="2">db &lt;-<span class="st"> </span>insect<span class="op">::</span><span class="kw">taxonomy</span>(<span class="dt">db =</span> <span class="st">&quot;NCBI&quot;</span>, <span class="dt">synonyms =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb6-3" title="3"></a>
<a class="sourceLine" id="cb6-4" title="4"><span class="co">#Filter the taxonomy database to remove contaminants and </span></a>
<a class="sourceLine" id="cb6-5" title="5">db &lt;-<span class="st"> </span>db <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-6" title="6"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span>rank <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;varietas&quot;</span>,<span class="st">&quot;subspecies&quot;</span>,<span class="st">&quot;species subgroup&quot;</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-7" title="7"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;sp.&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-8" title="8"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;spp.&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-9" title="9"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;aff.&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-10" title="10"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;nr.&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-11" title="11"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;bv.&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-12" title="12"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;cf.&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-13" title="13"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;nom.&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-14" title="14"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;nud.&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-15" title="15"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;environment&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-16" title="16"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;undescribed&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-17" title="17"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;unverified&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-18" title="18"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;unclassified&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-19" title="19"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;uncultured&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-20" title="20"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;unidentif&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-21" title="21"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;NA&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-22" title="22"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name, <span class="kw">fixed</span>(<span class="st">&quot;error&quot;</span>))) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-23" title="23"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name,<span class="st">&quot;[0-9]&quot;</span>))<span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-24" title="24"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">str_detect</span>(name,<span class="st">&quot;[:punct:]&quot;</span>))</a>
<a class="sourceLine" id="cb6-25" title="25"></a>
<a class="sourceLine" id="cb6-26" title="26">remove &lt;-<span class="st"> </span><span class="kw">names</span>(filtered)  <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-27" title="27"><span class="st">  </span><span class="kw">str_split_fixed</span>(<span class="st">&quot;;&quot;</span>, <span class="dt">n =</span> <span class="dv">2</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-28" title="28"><span class="st">  </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-29" title="29"><span class="st">  </span><span class="kw">filter</span>(V2 <span class="op">%in%</span><span class="st"> </span>db<span class="op">$</span>name) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-30" title="30"><span class="st">  </span><span class="kw">unite</span>(names,<span class="kw">c</span>(<span class="st">&quot;V1&quot;</span>,<span class="st">&quot;V2&quot;</span>),<span class="dt">sep=</span><span class="st">&quot;;&quot;</span>)</a>
<a class="sourceLine" id="cb6-31" title="31"></a>
<a class="sourceLine" id="cb6-32" title="32">subset &lt;-<span class="st"> </span>filtered[<span class="kw">names</span>(filtered) <span class="op">%in%</span><span class="st"> </span>remove<span class="op">$</span>names]</a>
<a class="sourceLine" id="cb6-33" title="33"></a>
<a class="sourceLine" id="cb6-34" title="34"><span class="co">#Save names into attributes</span></a>
<a class="sourceLine" id="cb6-35" title="35"><span class="kw">attributes</span>(filtered)<span class="op">$</span>oldnames &lt;-<span class="st"> </span><span class="kw">names</span>(filtered)</a>
<a class="sourceLine" id="cb6-36" title="36"></a>
<a class="sourceLine" id="cb6-37" title="37"><span class="co">#Transform names to format appropriate for insect::purge</span></a>
<a class="sourceLine" id="cb6-38" title="38"><span class="kw">names</span>(filtered) &lt;-<span class="st"> </span><span class="kw">names</span>(filtered) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-39" title="39"><span class="st">  </span><span class="kw">str_split_fixed</span>(<span class="st">&quot;;&quot;</span>,<span class="dt">n=</span><span class="dv">2</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-40" title="40"><span class="st">  </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-41" title="41"><span class="st">  </span><span class="kw">pull</span>(<span class="st">&quot;V1&quot;</span>) </a>
<a class="sourceLine" id="cb6-42" title="42"></a>
<a class="sourceLine" id="cb6-43" title="43"><span class="co">#Retain unique names only</span></a>
<a class="sourceLine" id="cb6-44" title="44">filtered &lt;-<span class="st"> </span>insect<span class="op">::</span><span class="kw">subset.DNAbin</span>(filtered, <span class="dt">subset =</span> <span class="op">!</span><span class="kw">duplicated</span>(<span class="kw">names</span>(filtered)))</a>
<a class="sourceLine" id="cb6-45" title="45"></a>
<a class="sourceLine" id="cb6-46" title="46"><span class="co">#Cluster and remove misannotated sequences</span></a>
<a class="sourceLine" id="cb6-47" title="47">purged  &lt;-<span class="st"> </span>insect<span class="op">::</span><span class="kw">purge</span>(filtered, <span class="dt">db =</span> db, <span class="dt">level =</span> <span class="st">&quot;species&quot;</span>, <span class="dt">confidence =</span> <span class="fl">0.8</span>,</a>
<a class="sourceLine" id="cb6-48" title="48">                  <span class="dt">threshold =</span> <span class="fl">0.99</span>, <span class="dt">method =</span> <span class="st">&quot;farthest&quot;</span>)</a>
<a class="sourceLine" id="cb6-49" title="49"></a>
<a class="sourceLine" id="cb6-50" title="50"><span class="co">#Restore old names</span></a>
<a class="sourceLine" id="cb6-51" title="51"><span class="kw">names</span>(purged) &lt;-<span class="st"> </span><span class="kw">attributes</span>(purged)<span class="op">$</span>oldnames</a></code></pre></div>
</div>
<div id="resolve-synonyms" class="section level2">
<h2>Resolve synonyms</h2>
<p>Classification of sequences into reference taxonomy can be complicated by the existence of taxonomic synonyms. To resolve this we use the GBIF server to check each name to see if it represents a currently valid taxa. If it represents a synonym, the name is replaced with the accepted taxon name.</p>
<p>The options to consider here, is what to do with the synonyms that dont exist in the NCBI taxonomy, in this case we ignore the fact taxa are missing from the NCBI taxonomy and rename them anyway</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">resolved &lt;-<span class="st"> </span><span class="kw">resolve_synonyms</span>(subset,<span class="dt">subspecies=</span><span class="ot">FALSE</span>,<span class="dt">quiet=</span><span class="ot">FALSE</span>,<span class="dt">missing=</span><span class="st">&quot;ignore&quot;</span>,<span class="dt">higherrank=</span><span class="ot">FALSE</span>,<span class="dt">fuzzy=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb7-2" title="2"></a>
<a class="sourceLine" id="cb7-3" title="3"><span class="co">#Check for differences in names</span></a>
<a class="sourceLine" id="cb7-4" title="4"><span class="kw">names</span>(resolved)[<span class="kw">which</span>(<span class="op">!</span><span class="kw">names</span>(resolved) <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(subset))]</a></code></pre></div>
</div>
<div id="prune-large-groups" class="section level2">
<h2>Prune large groups</h2>
<p>In many cases groups of taxa are over-represented in databases, which can slow down and in some cases bias the taxonomic assignment process. Here we prune over-represented groups down to 5 sequences. Here we have the option of discarding these sequences by length (ie removing smaller sequences first), or randomly.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="co">#Prune group sizes down to 5, removing all identical sequences first</span></a>
<a class="sourceLine" id="cb8-2" title="2">pruned &lt;-<span class="st"> </span><span class="kw">prune_groups</span>(resolved,<span class="dt">maxGroupSize =</span> <span class="dv">5</span>, <span class="dt">discardby=</span><span class="st">&quot;length&quot;</span>,<span class="dt">dedup=</span><span class="ot">TRUE</span>, <span class="dt">quiet =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
</div>
<div id="trim-to-primer-regions" class="section level2">
<h2>Trim to primer regions</h2>
<p>Next we will trim the sequences to the primer regions we use for metabarcoding using the virtualPCR function from the insect R pacakge</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="co">#Trim to primer region using virtualPCR from insect package</span></a>
<a class="sourceLine" id="cb9-2" title="2">amplicon &lt;-<span class="st"> </span>insect<span class="op">::</span><span class="kw">virtualPCR</span>(pruned, <span class="dt">up =</span> <span class="st">&quot;ACWGGWTGRACWGTNTAYCC&quot;</span>,<span class="dt">down=</span> <span class="st">&quot;ARYATDGTRATDGCHCCDGC&quot;</span>,<span class="dt">cores=</span><span class="dv">2</span>, <span class="dt">rcdown =</span> <span class="ot">TRUE</span>, <span class="dt">trimprimers =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="reformat-to-taxonomic-classifier" class="section level2">
<h2>Reformat to taxonomic classifier</h2>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="co">#Change to complete taxonomic heirarchy </span></a>
<a class="sourceLine" id="cb10-2" title="2">heirarchy &lt;-<span class="st"> </span><span class="kw">reformat_heirarchy</span>(amplicon, <span class="dt">db=</span>db, <span class="dt">quiet=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb10-3" title="3"></a>
<a class="sourceLine" id="cb10-4" title="4"><span class="co">#Reformat to Kingdom to genus heirarchy suitable for assigntaxonomy classifier in DADA2</span></a>
<a class="sourceLine" id="cb10-5" title="5">dada2_gen &lt;-<span class="st"> </span><span class="kw">reformat_dada2_gen</span>(amplicon, <span class="dt">db=</span>db, <span class="dt">quiet=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb10-6" title="6"></a>
<a class="sourceLine" id="cb10-7" title="7"><span class="co">#Reformat to genus species binomials as suitable for assignSpecies in DADA2</span></a>
<a class="sourceLine" id="cb10-8" title="8">dada2_spp &lt;-<span class="st"> </span><span class="kw">reformat_dada2_spp</span>(amplicon)</a></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
