<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Author: Alexander M Piper" />

<meta name="date" content="2020-07-17" />

<title>taxreturn</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">taxreturn</h1>
<h3 class="subtitle">An R package for download and curation of public DNA barcode reference data</h3>
<h4 class="author">Author: Alexander M Piper</h4>
<h4 class="date">2020-07-17</h4>



<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>As with conventional DNA barcoding, accurate taxonomic assignment in metabarcoding studies relies on a well-curated reference database of DNA marker sequences to compare query sequences against. The primary public nucleotide databases of relevance to eukaryotic metabarcoding are the NCBI GenBank database and the Barcode of Life Data System (BOLD), both of which taxreturn supports. While GenBank hosts greater overall sequence data, BOLD represents a more curated DNA barcoding database that aims to maintain consistent links between sequences, validated morphological specimens, and associated specimen collection metadata.</p>
<p>Despite the best efforts of submitters, both these databases have well documented issues with barcode sequences being either insufficiently annotated, annotated with the incorrect species, or multiple morpho-species assigned to the same DNA barcode, which may reflect misidentifications or the existence of species complexes.While some metabarcoding studies have responded to the aforementioned issues by exclusively using in-house reference databases for taxonomic assignment [90, 93–95], because many insect surveillance programmes aim to detect species that are not locally present, the reliance on public data to supplement in-house sequences may be unavoidable. Curating public reference data is therefore crucial for ensuring the robustness of a metabarcoding analysis, and this is the role that the taxreturn package aims to fulfil.</p>
<p>The main steps of the taxreturn workflow are as follows:</p>
<ul>
<li>Download relevant barcode sequences and Mitochondrial genomes from BOLD and Genbank</li>
<li>Use a profile hidden markov model trained on a reference alignment for the locus to remove non-homologous loci and cut the desired region from mitogenomes/longer sequences</li>
<li>Use sequence similarity clustering to remove misannotated sequences (ie 99% similar sequences should share species level taxononmy)</li>
<li>Resolve taxonomic synonyms using a reference taxonomy database (NCBI or open tree of life)</li>
<li>Prune sequences from over-represented groups</li>
<li>Reformat into a taxonomic hierarchy format (ie Kingdom;phylum….;Species) appropriate for metabarcoding classifiers.</li>
</ul>
</div>
<div id="a-worked-example" class="section level1">
<h1>A Worked example</h1>
<p>The following vignette will run through an example analysis of retrieving insect COI barcode sequences and mitochondrial genomes, curating these, and outputting a reference database.</p>
<div id="install-load-package" class="section level2">
<h2>Install &amp; load package</h2>
<p>First, make sure that the devtools, Biostrings and ape packages are installed and up to date. Then install and load the latest development version of the taxreturn package from GitHub as follows:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;alexpiper/taxreturn&quot;</span>)</a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw">library</span>(taxreturn)</a>
<a class="sourceLine" id="cb1-3" title="3"></a>
<a class="sourceLine" id="cb1-4" title="4"><span class="co"># Load other necessary packages</span></a>
<a class="sourceLine" id="cb1-5" title="5"><span class="kw">library</span>(Biostrings)</a>
<a class="sourceLine" id="cb1-6" title="6"><span class="kw">library</span>(ape)</a>
<a class="sourceLine" id="cb1-7" title="7"><span class="kw">library</span>(insect)</a>
<a class="sourceLine" id="cb1-8" title="8"><span class="kw">library</span>(tidyverse)</a></code></pre></div>
</div>
</div>
<div id="fetching-sequences-from-genbank-and-bold" class="section level1">
<h1>Fetching sequences from GenBank and BOLD</h1>
<p>The first step is to retrieve public reference data sequences from NCBI Genbank and BOLD. This can be done with the fetchSeqs function, which wraps an interface to entrez and BOLD API’s respectively. This function can either take a single higher level rank, i.e. the family ‘Trioza’, or it can take a vector of taxon names, such as the species contained on a priority pest list or those of conservation concern.</p>
<p>The downstream option lets the function know if you want to conduct searches and output fasta files using the input rank, or at a taxonomic rank downstream of it, i.e. Species. The downto option then determines what that downstream rank is. This functionality is important for getting around server limits when conducting large searches for example ‘Insecta’, however comes with some computational overhead when downloading few taxa.</p>
<p>The marker option determines what will be in the search, note that the naming of loci differs between Genbank and bold so i suggest conducting a test search on their respective websites to confirm the desired marker.</p>
<p>The compress option toggles whether to output as a gzipped fasta file to save space. All further functions in this package should be capable of handling gzipped files.</p>
<p>Finally, the cores option determines how many cores to use when downloading sequences, using more cores can speed up searches. <em>note - i dont believe cores is currently working properly so use 1 core for now</em></p>
<p>Depending on the amount of sequences you are downloading, and the speed of your internet connection this step can take from minutes to hours, i suggest running this overnight for large searches with over a million sequences.</p>
<div id="download-sequences-from-ncbi-genbank" class="section level2">
<h2>Download sequences from NCBI GenBank</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="co">## Fetch sequences from GenBank by searching for a taxon name</span></a>
<a class="sourceLine" id="cb2-2" title="2"><span class="kw">fetchSeqs</span>(<span class="st">&quot;Scaptodrosophila&quot;</span>, <span class="dt">database=</span><span class="st">&quot;genbank&quot;</span>, <span class="dt">out.dir=</span><span class="st">&quot;genbank&quot;</span>, <span class="dt">downstream=</span><span class="ot">FALSE</span>, <span class="dt">marker=</span><span class="st">&quot;COI OR COI OR COX1 OR COXI&quot;</span>, <span class="dt">quiet=</span><span class="ot">FALSE</span>, <span class="dt">output =</span> <span class="st">&quot;gb-binom&quot;</span>, <span class="dt">compress=</span><span class="ot">TRUE</span>, <span class="dt">force=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb2-3" title="3"></a>
<a class="sourceLine" id="cb2-4" title="4"><span class="co">## OR fetch sequences from a species list</span></a>
<a class="sourceLine" id="cb2-5" title="5">spp_list &lt;-<span class="st"> </span><span class="kw">readLines</span>(<span class="st">&quot;species_list.txt&quot;</span>)</a>
<a class="sourceLine" id="cb2-6" title="6"><span class="kw">fetchSeqs</span>(spp_list, <span class="dt">database=</span><span class="st">&quot;genbank&quot;</span>, <span class="dt">out.dir=</span><span class="st">&quot;genbank&quot;</span>, <span class="dt">quiet=</span><span class="ot">FALSE</span>, <span class="dt">output =</span> <span class="st">&quot;gb-binom&quot;</span>, <span class="dt">marker=</span><span class="st">&quot;COI OR COI OR COX1 OR COXI&quot;</span>, <span class="dt">compress=</span><span class="ot">TRUE</span>, <span class="dt">force=</span><span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="download-sequences-from-bold" class="section level2">
<h2>Download sequences from BOLD</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="co">## Fetch sequences from BOLD by searching for a taxon name</span></a>
<a class="sourceLine" id="cb3-2" title="2"><span class="kw">fetchSeqs</span>(<span class="st">&quot;Scaptodrosophila&quot;</span>, <span class="dt">database=</span><span class="st">&quot;bold&quot;</span>, <span class="dt">out.dir=</span><span class="st">&quot;bold&quot;</span>, <span class="dt">downstream=</span><span class="ot">FALSE</span>, <span class="dt">quiet=</span><span class="ot">FALSE</span>, <span class="dt">marker=</span><span class="st">&quot;COI-5P&quot;</span>, <span class="dt">output =</span> <span class="st">&quot;gb-binom&quot;</span>, <span class="dt">compress=</span><span class="ot">TRUE</span>, <span class="dt">force=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb3-3" title="3"></a>
<a class="sourceLine" id="cb3-4" title="4"><span class="co">## OR fetch sequences from a species list</span></a>
<a class="sourceLine" id="cb3-5" title="5">spp_list &lt;-<span class="st"> </span><span class="kw">readLines</span>(<span class="st">&quot;species_list.txt&quot;</span>)</a>
<a class="sourceLine" id="cb3-6" title="6"><span class="kw">fetchSeqs</span>(spp_list, <span class="dt">database=</span><span class="st">&quot;bold&quot;</span>, <span class="dt">out.dir=</span><span class="st">&quot;bold&quot;</span>, <span class="dt">quiet=</span><span class="ot">FALSE</span>, <span class="dt">marker=</span><span class="st">&quot;COI-5P&quot;</span>, <span class="dt">output =</span> <span class="st">&quot;gb-binom&quot;</span>, <span class="dt">compress=</span><span class="ot">TRUE</span>, <span class="dt">force=</span><span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="download-mitochondrial-genomes-from-genbank" class="section level2">
<h2>Download mitochondrial genomes from GenBank</h2>
<p>We can also use these functions to download mitochondrial genomes, and use the PHMM implemented in the clean_seqs function to pull out our target region from the genome.</p>
<p>The fetchSeqs function accepts a special input argument of marker=“mitochondria” to do this.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="co"># Fetch mitochondrial genomes from genbank by searching for a taxon name</span></a>
<a class="sourceLine" id="cb4-2" title="2"><span class="kw">fetchSeqs</span>(<span class="st">&quot;Scaptodrosophila&quot;</span>, <span class="dt">database=</span><span class="st">&quot;genbank&quot;</span>, <span class="dt">out.dir=</span><span class="st">&quot;genbank&quot;</span>, <span class="dt">quiet=</span><span class="ot">FALSE</span>, <span class="dt">marker=</span><span class="st">&quot;mitochondria&quot;</span>, <span class="dt">output =</span> <span class="st">&quot;gb-binom&quot;</span>, <span class="dt">compress=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb4-3" title="3"></a>
<a class="sourceLine" id="cb4-4" title="4"><span class="co">## OR fetch sequences from a species list</span></a>
<a class="sourceLine" id="cb4-5" title="5">spp_list &lt;-<span class="st"> </span><span class="kw">readLines</span>(<span class="st">&quot;species_list.txt&quot;</span>)</a>
<a class="sourceLine" id="cb4-6" title="6"><span class="kw">fetchSeqs</span>(spp_list, <span class="dt">database=</span><span class="st">&quot;genbank&quot;</span>, <span class="dt">out.dir=</span><span class="st">&quot;genbank&quot;</span>, <span class="dt">quiet=</span><span class="ot">FALSE</span>, <span class="dt">marker=</span><span class="st">&quot;mitochondria&quot;</span>, <span class="dt">output =</span> <span class="st">&quot;gb-binom&quot;</span>, <span class="dt">compress=</span><span class="ot">TRUE</span>)</a></code></pre></div>
</div>
</div>
<div id="curating-public-reference-sequences" class="section level1">
<h1>Curating public reference sequences</h1>
<p>Due to the aforementioned issues with misannotated sequences in public reference databases we will use a number of filtering steps.</p>
<div id="removing-non-homologous-sequences" class="section level2">
<h2>Removing non-homologous sequences</h2>
<p>The first step is to remove non-homologous loci by mapping to a reference of COI. In order to reduce bias involved in mapping sequences to a single reference sequence, we will align them to a profile hidden markov model of the gene that uses a probablistic framework to take into account a wider range of diversity. To make this model we will use the aphid R package, and a curated alignment of insect COI sequences obtained from the Midori dataset and trimmed to the folmer region.</p>
<p>A pretrained model of the conventional COI barcode or ‘folmer’ region can be loaded from the package data as below. This model was trained on the midori longest dataset of all COI sequences.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="co">#This loads the model into the workspace</span></a>
<a class="sourceLine" id="cb5-2" title="2"><span class="kw">data</span>(<span class="st">&quot;model&quot;</span>, <span class="dt">package=</span><span class="st">&quot;taxreturn&quot;</span>)</a>
<a class="sourceLine" id="cb5-3" title="3"></a>
<a class="sourceLine" id="cb5-4" title="4"><span class="co">#See what it looks like</span></a>
<a class="sourceLine" id="cb5-5" title="5"><span class="kw">print</span>(model)</a></code></pre></div>
<p>However if you are workign with a different barcode locus, or if you wish to improve accuracy by training on a specific taxonomic group you can train a new PHMM model using the aphid R package as below:</p>
<p>Note: if you are using a public dataset as a reference alignment to build the model, it may be worth further curating it manually first, as a poor reference alignment will produce a poor model and greatly affect downstream analysis</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="co"># Read in sequence dataset to be used in model training</span></a>
<a class="sourceLine" id="cb6-2" title="2">seqs &lt;-<span class="st">  </span><span class="kw">readDNAStringSet</span>(<span class="st">&quot;MIDORI_LONGEST_20180221_COI.fasta&quot;</span>)</a>
<a class="sourceLine" id="cb6-3" title="3"></a>
<a class="sourceLine" id="cb6-4" title="4"><span class="co"># Trim the sequences to the amplified region using a virtual PCR</span></a>
<a class="sourceLine" id="cb6-5" title="5">amplicon &lt;-<span class="st"> </span>insect<span class="op">::</span><span class="kw">virtualPCR</span>(seqs,</a>
<a class="sourceLine" id="cb6-6" title="6">                               <span class="dt">up =</span> <span class="st">&quot;TITCIACIAAYCAYAARGAYATTGG&quot;</span>,  <span class="co">#Forward primer</span></a>
<a class="sourceLine" id="cb6-7" title="7">                               <span class="dt">down=</span> <span class="st">&quot;TAIACYTCIGGRTGICCRAARAAYCA&quot;</span>, <span class="co">#Reverse primer</span></a>
<a class="sourceLine" id="cb6-8" title="8">                             <span class="dt">cores=</span><span class="dv">1</span>, <span class="dt">rcdown =</span> <span class="ot">TRUE</span>, <span class="dt">trimprimers =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb6-9" title="9"></a>
<a class="sourceLine" id="cb6-10" title="10"><span class="co">#Only retain amplicons of the appropriate length (in this case 658bp)</span></a>
<a class="sourceLine" id="cb6-11" title="11">amplicon_filtered &lt;-<span class="st"> </span>amplicon[<span class="kw">lengths</span>(amplicon) <span class="op">==</span><span class="st"> </span><span class="dv">658</span>]</a>
<a class="sourceLine" id="cb6-12" title="12">model &lt;-<span class="st"> </span>aphid<span class="op">::</span><span class="kw">derivePHMM</span>(amplicon_filtered)</a></code></pre></div>
<p>Now that we have our train PHMM model, we will use the cleanseqs function with the model to remove non-homolgous sequences and extract the target locus from mitochondrial genomes and longer sequences.</p>
<p>Firstly we will merge the sequences from each database together. As BOLD and GenBank share a number of sequences, we subset the merged file to only the unique sequences to speed things up.</p>
<p>As we only wish to look at the sequence data contained within the range of our alignment model (in this case the folmer region of COI), we use the option shave=TRUE to remove all bases to the left and right of the model.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"><span class="co">#read in all fastas and merge</span></a>
<a class="sourceLine" id="cb7-2" title="2">seqs &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">readDNAStringSet</span>(<span class="kw">list.files</span>(<span class="st">&quot;genbank&quot;</span>, <span class="dt">pattern =</span> <span class="st">&quot;.fa&quot;</span>,</a>
<a class="sourceLine" id="cb7-3" title="3">                                      <span class="dt">full.names =</span> <span class="ot">TRUE</span>)),</a>
<a class="sourceLine" id="cb7-4" title="4">                <span class="kw">readDNAStringSet</span>(<span class="kw">list.files</span>(<span class="st">&quot;bold&quot;</span>, <span class="dt">pattern =</span> <span class="st">&quot;.fa&quot;</span>,</a>
<a class="sourceLine" id="cb7-5" title="5">                                            <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb7-6" title="6">                )</a>
<a class="sourceLine" id="cb7-7" title="7"></a>
<a class="sourceLine" id="cb7-8" title="8"><span class="co"># Remove those sequnce accessions that are identical across both databases</span></a>
<a class="sourceLine" id="cb7-9" title="9"><span class="co"># Using a regex that splits to just the accessions</span></a>
<a class="sourceLine" id="cb7-10" title="10">uniqSeqs &lt;-<span class="st"> </span>seqs[<span class="op">!</span><span class="kw">duplicated</span>(<span class="kw">str_extract</span>(<span class="kw">names</span>(seqs), <span class="st">&quot;^.*</span><span class="ch">\\</span><span class="st">|&quot;</span> )),] </a>
<a class="sourceLine" id="cb7-11" title="11"></a>
<a class="sourceLine" id="cb7-12" title="12"><span class="co">#remove non-homologous sequences</span></a>
<a class="sourceLine" id="cb7-13" title="13">filtered &lt;-<span class="st"> </span><span class="kw">clean_seqs</span>(uniqSeqs, model, <span class="dt">minscore =</span> <span class="dv">100</span>, <span class="dt">cores=</span><span class="dv">1</span>, <span class="dt">shave=</span><span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="resolve-contaminated-sequences-and-misannotated-taxonomy" class="section level2">
<h2>Resolve Contaminated sequences and misannotated taxonomy</h2>
<p>The other main form of misannotation that can effect metabarcoding datasets is incorrect taxonomy for the reference sequences. To resolve this issue, we will cluster sequences at a specific similarity threshold (in this case 97% simularity), and compares the heirarchial taxonomy within clusters. When the taxonomy of a sequences diverges from the other sequences in its cluster, it is removed as a putative misannotation. The confidence required the remove a sequence can be adjusted. In this case we use a confidence threshold of 0.8, which indicates the putative misannotated sequence must diverge from 4/5 other sequences in its cluster to be removed from the dataset.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="co"># Load the NCBI taxonomic lineage</span></a>
<a class="sourceLine" id="cb8-2" title="2">db &lt;-<span class="st"> </span><span class="kw">get_ncbi_lineage</span>()</a>
<a class="sourceLine" id="cb8-3" title="3"></a>
<a class="sourceLine" id="cb8-4" title="4"><span class="co">#Get db</span></a>
<a class="sourceLine" id="cb8-5" title="5">mixed_clusters &lt;-<span class="st"> </span>taxreturn<span class="op">::</span><span class="kw">get_mixed_clusters</span>(</a>
<a class="sourceLine" id="cb8-6" title="6">    <span class="dt">x =</span> filtered, <span class="dt">db=</span>db,</a>
<a class="sourceLine" id="cb8-7" title="7">    <span class="dt">rank =</span> <span class="st">&quot;genus&quot;</span>,</a>
<a class="sourceLine" id="cb8-8" title="8">    <span class="dt">threshold =</span> <span class="fl">0.97</span>,</a>
<a class="sourceLine" id="cb8-9" title="9">    <span class="dt">return =</span> <span class="st">&quot;consensus&quot;</span>,</a>
<a class="sourceLine" id="cb8-10" title="10">    <span class="dt">confidence=</span><span class="fl">0.6</span>, <span class="dt">quiet =</span> <span class="ot">FALSE</span>) </a>
<a class="sourceLine" id="cb8-11" title="11"></a>
<a class="sourceLine" id="cb8-12" title="12"><span class="co"># Get accession numbers to remove</span></a>
<a class="sourceLine" id="cb8-13" title="13">rem &lt;-<span class="st"> </span>mixed_clusters<span class="op">$</span>Acc</a>
<a class="sourceLine" id="cb8-14" title="14"></a>
<a class="sourceLine" id="cb8-15" title="15">purged  &lt;-<span class="st"> </span><span class="kw">subset.DNAbin</span>(filtered, <span class="dt">subset =</span> <span class="op">!</span><span class="kw">str_replace</span>(<span class="kw">names</span>(filtered), <span class="st">&quot;(?:.(?!;))+$&quot;</span>, <span class="st">&quot;&quot;</span>) <span class="op">%in%</span><span class="st"> </span>rem)</a></code></pre></div>
</div>
<div id="optional-blast-against-wolbachia-genome" class="section level2">
<h2>Optional: Blast against wolbachia genome</h2>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="kw">fetchSeqs</span>(<span class="st">&quot;wolbachia pipientis&quot;</span>, <span class="dt">database=</span><span class="st">&quot;genbank&quot;</span>, <span class="dt">out.dir=</span><span class="st">&quot;wolbachia&quot;</span>, <span class="dt">quiet=</span><span class="ot">FALSE</span>, <span class="dt">marker=</span><span class="st">&quot;genome&quot;</span>, <span class="dt">output =</span> <span class="st">&quot;gb-binom&quot;</span>, <span class="dt">compress=</span><span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="optional-filter-for-stop-codons" class="section level2">
<h2>Optional: Filter for stop codons</h2>
<p>As the COI barcode is a protein coding region, underlying evolutionary constraints mean that any sequences containing stop codons or indels of lengths which are not a multiple of 3 commonly indicate pseudogenes.</p>
<p>The below function will remove any sequences containing these patterns. If you are not working with a protein coding gene, skip this section.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">codon_filt &lt;-<span class="st"> </span><span class="kw">codon_filter</span>(purged)</a></code></pre></div>
</div>
<div id="resolve-synonyms" class="section level2">
<h2>Resolve synonyms</h2>
<p>Classification of sequences into reference taxonomy can be complicated by the existence of taxonomic synonyms. To resolve this we use the NCBI names file to check each name to see if it represents a currently valid taxa. If it represents a synonym, the name is replaced with the accepted taxon name.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">resolved &lt;-<span class="st"> </span><span class="kw">resolve_synonyms_ncbi</span>(purged)</a>
<a class="sourceLine" id="cb11-2" title="2"></a>
<a class="sourceLine" id="cb11-3" title="3"><span class="co">#Check for differences in names</span></a>
<a class="sourceLine" id="cb11-4" title="4"><span class="kw">names</span>(resolved)[<span class="kw">which</span>(<span class="op">!</span><span class="kw">names</span>(resolved) <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(purged))]</a></code></pre></div>
</div>
<div id="prune-large-groups" class="section level2">
<h2>Prune large groups</h2>
<p>In many cases groups of taxa are over-represented in databases, which can slow down and bias the taxonomic assignment process. Here we prune over-represented groups down to 5 sequences. This function has the option of discarding these sequences by length (ie removing smaller sequences first), or randomly.</p>
<p>We can also choose to de-duplicate the sequences here, which will remove all completely identical sequences before pruning down the group sizes.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1"><span class="co">#Prune group sizes down to 5, removing all identical sequences first</span></a>
<a class="sourceLine" id="cb12-2" title="2">pruned &lt;-<span class="st"> </span><span class="kw">prune_groups</span>(resolved, <span class="dt">maxGroupSize =</span> <span class="dv">5</span>, <span class="dt">discardby=</span><span class="st">&quot;length&quot;</span>, <span class="dt">dedup=</span><span class="ot">TRUE</span>, <span class="dt">quiet =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
</div>
<div id="optional-trim-to-primer-regions" class="section level2">
<h2>Optional: Trim to primer regions</h2>
<p>Depending on the taxonomic classifier used, it can be beneficial to trim the reference sequences down to the actual region amplified by the primers. Below we will trim the sequences to the primer regions we use for metabarcoding using the virtualPCR function from the insect R pacakge.</p>
<p>However, if you wish to keep the full length sequences, skip this section.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1"><span class="co">#Trim to primer region using virtualPCR from insect package</span></a>
<a class="sourceLine" id="cb13-2" title="2">amplicon &lt;-<span class="st"> </span>insect<span class="op">::</span><span class="kw">virtualPCR</span>(pruned, <span class="dt">up =</span> <span class="st">&quot;ACWGGWTGRACWGTNTAYCC&quot;</span>, <span class="dt">down=</span> <span class="st">&quot;ARYATDGTRATDGCHCCDGC&quot;</span>, <span class="dt">cores=</span><span class="dv">1</span>, <span class="dt">rcdown =</span> <span class="ot">TRUE</span>, <span class="dt">trimprimers =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb13-3" title="3"></a>
<a class="sourceLine" id="cb13-4" title="4"><span class="co">#Check the lengths of the trimmed sequences</span></a>
<a class="sourceLine" id="cb13-5" title="5"><span class="kw">table</span>(<span class="kw">lengths</span>(amplicon))</a></code></pre></div>
</div>
<div id="reformat-to-taxonomic-hierarchy" class="section level2">
<h2>Reformat to taxonomic hierarchy</h2>
<p>Finally, we will reformat to fasta annotations to contain the full taxonomic hierarchy and write out the fasta</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1"><span class="co">#Load NCBI taxonomy</span></a>
<a class="sourceLine" id="cb14-2" title="2">db &lt;-<span class="st"> </span><span class="kw">get_ncbi_lineage</span>()</a>
<a class="sourceLine" id="cb14-3" title="3"></a>
<a class="sourceLine" id="cb14-4" title="4"><span class="co">#Reformat to complete taxonomic heirarchy </span></a>
<a class="sourceLine" id="cb14-5" title="5">hierarchy &lt;-<span class="st"> </span><span class="kw">reformat_hierarchy</span>(amplicon, <span class="dt">db=</span>db, <span class="dt">quiet=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb14-6" title="6"></a>
<a class="sourceLine" id="cb14-7" title="7"><span class="co">#See if this worked</span></a>
<a class="sourceLine" id="cb14-8" title="8"><span class="kw">head</span>(<span class="kw">names</span>(hierarchy))</a>
<a class="sourceLine" id="cb14-9" title="9"></a>
<a class="sourceLine" id="cb14-10" title="10"><span class="co"># Save a zipped fasta file of curated reference database</span></a>
<a class="sourceLine" id="cb14-11" title="11">insect<span class="op">::</span><span class="kw">writeFASTA</span>(hierarchy, <span class="dt">file =</span> <span class="st">&quot;COI_reference_hierarchial.fa.gz&quot;</span>, <span class="dt">compress =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="summarise-number-of-taxa-in-database" class="section level2">
<h2>Summarise number of taxa in database</h2>
<p>Finally, we can output a few summary metrics for our final curated database, such as the number of unique sequences for each taxonomic rank:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1"><span class="kw">names</span>(hierarchy) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb15-2" title="2"><span class="st">  </span><span class="kw">str_split_fixed</span>(<span class="st">&quot;;&quot;</span>, <span class="dt">n=</span><span class="dv">8</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb15-3" title="3"><span class="st">  </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb15-4" title="4"><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">separate</span>(V1, <span class="dt">into=</span><span class="kw">c</span>(<span class="st">&quot;Sequences&quot;</span>, <span class="st">&quot;taxids&quot;</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb15-5" title="5"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw">set_colnames</span>(<span class="kw">c</span>(<span class="st">&quot;Sequences&quot;</span>, <span class="st">&quot;tax_ids&quot;</span>, <span class="st">&quot;kingdom&quot;</span>, <span class="st">&quot;phylum&quot;</span>, <span class="st">&quot;class&quot;</span>, <span class="st">&quot;order&quot;</span>, <span class="st">&quot;family&quot;</span>, <span class="st">&quot;genus&quot;</span>, <span class="st">&quot;species&quot;</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb15-6" title="6"><span class="st">  </span><span class="kw">summarise_all</span>(n_distinct)</a></code></pre></div>
<p>We can also generate a tree like representation of the taxonomy which we can output as a newick file for analysis in other software</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1">tree &lt;-<span class="st"> </span><span class="kw">tax2tree</span>(hierarchy, <span class="dt">output=</span><span class="st">&quot;phylo&quot;</span>)</a>
<a class="sourceLine" id="cb16-2" title="2"><span class="kw">write.tree</span>(tree, <span class="st">&quot;tree.nwk&quot;</span>)</a></code></pre></div>
</div>
<div id="bonus-output-trained-classifiers" class="section level2">
<h2>Bonus: output trained classifiers</h2>
<p>The taxreturn package supports reformatting of the reference database to work with</p>
<p>The classifiers currently supported are:</p>
<ul>
<li>RDP classifier as implemented in DADA2 assignTaxonomy function</li>
<li>Exact matchign as implemented in DADA2 assignSpecies function</li>
<li>IDTAXA Classifier as implemented in the DECIPHER R package</li>
</ul>
<div id="rdp" class="section level3">
<h3>RDP</h3>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1"><span class="co">#Load NCBI taxonomy</span></a>
<a class="sourceLine" id="cb17-2" title="2">db &lt;-<span class="st"> </span><span class="kw">get_ncbi_lineage</span>()</a>
<a class="sourceLine" id="cb17-3" title="3"></a>
<a class="sourceLine" id="cb17-4" title="4"><span class="co">#Reformat to Kingdom to genus heirarchy suitable for assigntaxonomy classifier in DADA2</span></a>
<a class="sourceLine" id="cb17-5" title="5">dada2_gen &lt;-<span class="st"> </span><span class="kw">reformat_dada2_gen</span>(amplicon, <span class="dt">db=</span>db, <span class="dt">quiet=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb17-6" title="6">insect<span class="op">::</span><span class="kw">writeFASTA</span>(dada2_gen, <span class="dt">file =</span> <span class="st">&quot;COI_reference_dada2gen.fa.gz&quot;</span>, <span class="dt">compress =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb17-7" title="7"></a>
<a class="sourceLine" id="cb17-8" title="8"><span class="co">#Reformat to genus species binomials as suitable for assignSpecies in DADA2</span></a>
<a class="sourceLine" id="cb17-9" title="9">dada2_spp &lt;-<span class="st"> </span><span class="kw">reformat_dada2_spp</span>(amplicon)</a>
<a class="sourceLine" id="cb17-10" title="10">insect<span class="op">::</span><span class="kw">writeFASTA</span>(dada2_spp, <span class="dt">file =</span> <span class="st">&quot;COI_reference_dada2spp.fa.gz&quot;</span>, <span class="dt">compress =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="idtaxa" class="section level3">
<h3>IDTAXA</h3>
<p>Taxreturn implements a wrapper function for training the IDTAXA classifier, as below:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1"><span class="co">#Load NCBI taxonomy</span></a>
<a class="sourceLine" id="cb18-2" title="2">db &lt;-<span class="st"> </span><span class="kw">get_ncbi_lineage</span>()</a>
<a class="sourceLine" id="cb18-3" title="3"></a>
<a class="sourceLine" id="cb18-4" title="4"><span class="co"># Train IDTAXA</span></a>
<a class="sourceLine" id="cb18-5" title="5">trainingSet &lt;-<span class="st"> </span><span class="kw">train_idtaxa</span>(amplicon, <span class="dt">maxGroupSize=</span><span class="dv">10</span>, <span class="dt">maxIterations =</span> <span class="dv">3</span>,  <span class="dt">allowGroupRemoval =</span> <span class="ot">TRUE</span>,  <span class="dt">get_lineage=</span><span class="ot">TRUE</span>, <span class="dt">db =</span> db, <span class="dt">quiet =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb18-6" title="6"></a>
<a class="sourceLine" id="cb18-7" title="7"><span class="co">#Write out training set</span></a>
<a class="sourceLine" id="cb18-8" title="8"><span class="kw">saveRDS</span>(trainingSet, <span class="dt">file=</span><span class="st">&quot;trained_idtaxa.rds&quot;</span>)</a></code></pre></div>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
